src/src/PluginStructs/P145_data_struct.h

```diff
@@ -0,0 +1,104 @@
+// ----------------------------------------------------------------------------
+// P145 "Gases - MQxxx (MQ135 CO2, MQ3 Alcohol)"
+// Definition of sensor abstraction
+// See _P145_MQxxx.ino
+// 2023 By flashmark
+// ----------------------------------------------------------------------------
+#ifndef PLUGINSTRUCTS_P145_DATA_STRUCT_H
+#define PLUGINSTRUCTS_P145_DATA_STRUCT_H
+
+
+#include "../../_Plugin_Helper.h"
+#ifdef USES_P145
+
+// Maximum number of types that can be selected from
+#define P145_MAXTYPES  10
+
+// Analog input pin for sensor value
+#define P145_SENSOR_PIN A0
+
+// Supply voltage applied to the sensor
+#define P145_VCC        (5.0)
+// Max input scale voltage of the ADC
+#define P145_VMAX       (3.3)
+
+// Calibration interval in milli seconds (once per day)
+#define P145_CALIBRATION_INTERVAL (24*60*60*1000)
+
+// Selection of the conversion algorithm Rs/R0 => value
+enum P145_algorithm 
+{
+  p145AlgNone,  // Undefined/free algorithm
+  p145AlgA,     // MQ-135 with optional temp/hum correction 
+  p145AlgB,     // miguel5612 Exponential
+  p145AlgC      // miguel5612 Linear
+};
+
+// Structure with sensor specific conversion data
+struct P145_SENSORDEF
+{
+  char name[8];       // Sensor type name
+  char gas[8];        // Measured gas concentration
+  P145_algorithm alg; // Preferred/tuned algorithm
+  float cleanRatio;   // Rs/R0 ratio in clean air
+  float para;         // PARA scaling factor value
+  float parb;         // PARB exponent value
+  float cora;         // CORA
+  float corb;         // CORB
+  float corc;         // CORC
+  float cord;         // CORD
+  float core;         // CORE
+  float corf;         // CORF
+  float corg;         // CORG
+};
```

Comment:
Please sort the members of this class with the "largest" types first.
`float` is 4 byte-aligned, char-array is 1 byte aligned and I have no clue what the enum will be.
That's the real problem, that the alignment will be compiler dependent and thus you might get really funky issues when such struct data is stored as PROGMEM.
You can force a type on the enum, like this: `enum P145_SENSORDEF : uint32_t`
I know this is way too large for what is needed, but this way you will have a multiple of 4 bytes as struct size, which will probably cause no issues when reading it from PROGMEM.

Anyway, it is always a good idea to sort members based on their size as it will lead to the smallest possible structs in memory.

Reply:
Real embedded thinking :-) I really like such optimizations when required, as software architect I prefer a "logical" schema, especially as others will append the table with sensor types. Strings were deliberately set 8 charecters (or 16 if 8 would not fit). I will rearrane this as the table could become big. Although there will be one copy in ram. I don't know how PROGMEM data works. If possible the P145_data_struct code could access this static data through a pointer directly in PROGMEM. For now I will rearrange it.

Reply:
PROGMEM data is read per 32 bits at once, so better to have it all in multiples of 4 bytes.
Alignment may be an issue with PROGMEM stored structs. Therefore I rather have the structs sized at multiples of 4 bytes.
If you don't sort the members based on size, you may end up with either a larger struct, or the compiler may decide to make it a packed struct. Then the typically 4-byte types may no longer be 4-byte aligned and then you get all kinds of nasty things.
So better take extra care when using PROGMEM storage. Sometimes you need to read it first into regular memory using those `_P` postfixed functions and then map it to a struct or something.